src/main/java/edu/harvard/iq/dataverse/engine/command/impl/RestrictFileCommand.java

```diff
@@ -50,56 +50,41 @@ protected void executeImpl(CommandContext ctxt) throws CommandException {
+        // check if this file is already restricted or already unrestricted
+        if (restrict == file.getFileMetadata().isRestricted()) {
+            String text = restrict ? "restricted" : "unrestricted";
+            throw new CommandExecutionException("File " + file.getDisplayName() + " is already " + text, this);
+        }
+        // At present 4.9.4, it doesn't appear that new files use this command, so owner should always be set...
+        if (file.getOwner() == null) {
             // this is a new file through upload, restrict
             file.getFileMetadata().setRestricted(restrict);
-            file.setRestricted(restrict); 
+            file.setRestricted(restrict);
         }
-        
         else {
             Dataset dataset = file.getOwner();
             DatasetVersion workingVersion = dataset.getEditVersion();
-
-            // check if this file is already restricted or already unrestricted
-            if ((restrict && file.getFileMetadata().isRestricted()) || (!restrict  && !file.getFileMetadata().isRestricted())) {
-                String text =  restrict ? "restricted" : "unrestricted";
-                throw new CommandExecutionException("File " + file.getDisplayName() + " is already " + text, this);
-            }
-
-            // check if this dataset is a draft (should be), then we can update restrict
-            if (workingVersion.isDraft()) {
-                // required for updating from a published version to a draft
-                // because we must update the working version metadata
-                if (dataset.isReleased()){
-                    for (FileMetadata fmw : workingVersion.getFileMetadatas()) {
-                        if (file.equals(fmw.getDataFile())) {
-                            fmw.setRestricted(restrict);
-                            if (!file.isReleased()) {
-                                file.setRestricted(restrict); 
-                            }
-
-                        }
-
-                    }
-                }
-                else {
-                    file.getFileMetadata().setRestricted(restrict);
-                    if (!file.isReleased()) {
-                        file.setRestricted(restrict);
-                    }
-                    if (file.getFileMetadata().isRestricted() != restrict) {
-                        throw new CommandExecutionException("Failed to update the file metadata", this);
+            // We need the FileMetadata for the file in the draft dataset version and the
+            // file we have may still reference the fmd from the prior released version
+            FileMetadata draftFmd = file.getFileMetadata();
+            if (dataset.isReleased()) {
+                // We want to update the draft version, which may not exist (if the file has
+                // been deleted from an existing draft, so we want null unless this file's
+                // metadata can be found in the current version
+                draftFmd=null;
```

Comment:
This check here handles the dropped file case that was one reason for the double loop in DatasetPage